config/config.example.yml

```diff
@@ -42,26 +44,67 @@ services:
     priority: 1
     workspace_dir: ${WORKSPACE_DIR}
 
-# [Required] This section defines how to synchronize permissions between services when they shared resources
+# [Required] This section defines how to synchronize permissions between services when they share resources
 sync_permissions:
   # Friendly name to identify a sync point (The value is not used by Cowbird so this can be any relevant keyword)
   user_workspace:
     # [Required] This section defines a list of services (referring to the service name of the 'services' section)
     # sharing resources
-    services:
-      Geoserver: /api/workspaces/private/                # <Service name>:<Resource root matching with other services>
-      Thredds: /catalog/birdhouse/workspaces/private/    # Every service sharing the same resources can be added in this
-                                                         # list
+    services: # Contains the different resources that can be synchronized, ordered by service
+      Geoserver: # Service name, which should also be found in the services list at the top of this file
+        # Resource key (ex.: Geoserver_workspace): Custom name to represent a resource path.
+        # It should correspond to the names used in the `permission_mappings` section below.
+        # It contains the list of each segment of its resource path, with their corresponding name and type.
+        # The name of a segment can either contain a string name, a `SINGLE_TOKEN` `*` (to represent
+        # a single repetition of any name with the corresponding type) or a `MULTI_TOKEN` `**` (to
+        # represent any number (0 or more) of names that fit with the corresponding type). Note that tokens can
+        # only be found at the end of the path, so no named string should follow a token.
+        # Also, a MULTI_TOKEN can only be used one time in each list of path segments.
+        #
+        # The tokens are useful to know the types of any `suffix` path. A `suffix` path represents the segments
+        # at the end of a path that are not named in the config below. For example, with the config below, if we have
+        # an input resource path `/api/workspaces/private/dir1/dir2/file` that matches with the `Geoserver_workspace`
+        # resource key, we have a `suffix` path `/dir1/dir2/file`.
```

Comment:
Mmhm, maybe the example is more confusing than anything. 
In my example, the suffix path `/dir1/dir2/file` would match with the `**` token of `Geoserver_workspace`, only if all 3 segments of that path would be of `workspace` type. 

Then with the mappings, it is found that `Thredds_workspace` has to be synchronized. It checks if the suffix `/dir1/dir2/file` matches with the tokenized part of `Thredds_workspace`.
Tokenized part consists in one `**` and one `*`. Since the suffix has 3 segments, it can fit the tokens, using 2 `folder` types and 1 `file` type.

I guess a confusing part of this is that a segment named `file` is of the type `workspace` in `Geoserver_workspace`, but my intention was mostly to give an example that included resources from 2 different resources, and that showed the used of the different tokens.

Maybe I can add more examples, and remove the `file` part of this existing example to avoid confusions? Does this explanation help you?

Reply:
I'm mostly concerned by the mismatch between the final items on GeoServer side referring to workspace resources that somehow must be mapped to a file in Thredds. I wonder if everything can actually validated/created correctly. 

For example, let's assume that some resource `/api/workspaces/private/dir1/dir2/something` triggers a perm-sync from GeoServer towards Thredds. Let's assume also that Thredds currently only has `dir1` and `dir2` actually created as `directory`-typed resources in Magpie. There is no way to validate if `something` is a `file`-typed resource since it does not exist (yet). If the matching operation just assumes it is a `file` to fulfill the criteria, and the resource/permissions are applied for `something`, an error can happen. If later we receive another sync, `/api/workspaces/private/dir1/dir2/something/else`, then it still matches `workspace` types in GeoServer. The sync fails though on Thredds, since `something` that should be `directory` to accept a child `file` was previously enforced as `file` itself. 

Therefore, mixing distinct types and `**`/`*` simultaneously between services is bound to cause issues at some point.

Reply:
Yes, it's true that this example you gave could happen. It's challenging to find a solution that covers every possibility here, since there seems to be a lot of different cases, specific to each service. The problem is that this suffix part has to have a way to know its associated resource_types. For example, it has to know whether a `Thredds` resource is a `dir` or a `file` and we have to find a way to specify it in the config.

Not sure what would be best here. Either the user has to know and never use any of theses problematic cases in the config (for example, only map Geoserver `workspaces` with Thredds `dir`s and never to a `file`. But this is probably too risky to rely on just the user's knowledge, the user will probably not think of it and errors would happen. And there is probably many more other cases other than this specific one.

I'm wondering if we should just re-add the constraint that resources should already exist? That was part of the initial design anyway, I just removed that constraint since the `PATCH /permissions` was implemented to create resources automatically if they didn't exist. With this constraint, we would at least avoid having an error like this. I think `types` would not be necessary anymore in the config and we would go back to something closer to the initial solution, not needing these complicated tokens.

We would just use names in the resources config definition, and we would still use only the longest matching path. We wouldn't have to determine the types from the config, since the resources would already exist. An error would be thrown if any concerned resources would not exist.
Only thing is I'm back with the problem of finding a `resource_id` from a `resource_path` for the target permissions to synchronize, because the `PATCH /permissions` needs both resource names and types, and to get the types, I would have to get the resource tree, which requires the `resource_id`. 
Either that, or I could try to change `PATCH /permissions` to have a new 'mode' which only requires the resource_names, does not create any resources and instead checks that all resources already exist. Would have to check if this would work.

@fmigneault Are permissions specific to a service (and all its resources) or specific to a resource_type? Because if they are specific to `resource_types`, we might still have a problem, for example, with a suffix `/something/else`. If the `else` could be either a `dir` or a `file`, and the permissions to synchronize in the config are specific to the `dir` resource_type only.

@dbyrns @fmigneault Is there any reason not to use the constraint of checking for existing resources? I would need your input for this design. Or maybe there is a solution to the discussed problem here, without adding the constrating of using only existing resources...

Reply:
> For example, let's assume that some resource /api/workspaces/private/dir1/dir2/something triggers a perm-sync from GeoServer towards Thredds. Let's assume also that Thredds currently only has dir1 and dir2 actually created as directory-typed resources in Magpie. There is no way to validate if something is a file-typed resource since it does not exist (yet).

I agree with this specific relation. I think that the config could be validated somehow so that a relation cannot yield to multiple answers. Mapping both `file` and `dir` from thredds to a unique `workspace` type in geoserver yield to uncertainties. Removing the `file` part from the thredds side would remove any uncertainties. If a `workspace` is created then we create a `dir`. If a `dir` is created, then we create a `workspace`. The `**` and `*` can still be used to express relations but under some constraints.

In this specific case, does a `file` really need to be mapped to a `workspace`? Shouldn't it be a `layer`? If we had `**:workspace/*:layer <-> **:dir/*:file`, we would still get a valid relation that cannot yield uncertainties.

Reply:
> user has to know and never use any of theses problematic cases in the config

I think this is a viable option. The only problem I saw was that the provided example was already ambiguous and possibly proposing a config that could cause problems. If the example is adjusted to have valid cases, and some kind of warning to take these possible problem cases in consideration, it is fine. 

We don't have to solve all possible combinations, and we will be the ones making those configs anyway, so we can trust they should be done reasonably as well. I mostly want to make sure that a copy-paste of such example config is not prone to errors from the start, and that a reference reminds us after many months of not playing with this config that this case must be considered.

I think the constraint of existing resources is not good because Cowbird should more often then not have to deal with resources that were just created (on one service), and need sync on the other service side. We want to limit resource creation to the strict minimum (ie when a permission must be applied for one) to avoid slow permission resolutions. 

Permissions are service AND resource type specific. For example, `ServiceWFS` (plain) does not have any `workspace` concept, this is specific to `ServiceGeoserverWFS`. A `layer` is used in both though, but plain `ServiceWFS` does not allow `GetPropertyValue` permission.